tests/baselines/reference/genericMappedTypeAcceptsInferredObjectType.errors.txt

```diff
@@ -0,0 +1,64 @@
+
+
+==== tests/cases/compiler/genericMappedTypeAcceptsInferredObjectType.ts (6 errors) ====
+    interface Constraint {
+    	foo: string
+    }
+    
+    interface ExtendedConstraint extends Constraint {
+    	bar: string
+    }
+    
+    type GenericMap<T> = {
+    	[P in keyof T]: T[P]
+    }
+    
+    type OptionalGenericMap<T> = {
+    	[P in keyof T]?: T[P]
+    }
+    
+    const required = <T>(x: GenericMap<T>) => x
+    const optional = <T>(x: OptionalGenericMap<T>) => x
+    
+    const withinConstraint = <T extends Constraint>(foo: T['foo']) => {
+    	required<Constraint>({ foo }) // no error as { foo: T['foo'] } <: GenericMap<Constraint>
+    	required<T>({ foo }) // error as { foo: T['foo'] } /<: GenericMap<T> because other properties may be missing
+    	            ~~~~~~~
+!!! error TS2345: Argument of type '{ foo: T["foo"]; }' is not assignable to parameter of type 'GenericMap<T>'.
+    	optional<T>({}) // no error as {} <: OptionalGenericMap<Constraint>
+    	optional<T>({ foo }) // no error as { foo: T['foo'] } <: OptionalGenericMap<T>
+    }
+    
+    const withinExtendedConstraint = <T extends ExtendedConstraint>(foo: T['foo'], bar: T['bar']) => {
+    	required<ExtendedConstraint>({ foo }) // error as { foo: T['foo'] } /<: GenericMap<ExtendedConstraint> because bar is missing
+    	                             ~~~~~~~
+!!! error TS2345: Argument of type '{ foo: T["foo"]; }' is not assignable to parameter of type 'GenericMap<ExtendedConstraint>'.
+!!! error TS2345:   Property 'bar' is missing in type '{ foo: T["foo"]; }' but required in type 'GenericMap<ExtendedConstraint>'.
+!!! related TS2728 tests/cases/compiler/genericMappedTypeAcceptsInferredObjectType.ts:6:2: 'bar' is declared here.
+    	required<ExtendedConstraint>({ bar }) // error as { bar: T['bar'] } /<: GenericMap<ExtendedConstraint> because foo is missing
+    	                             ~~~~~~~
+!!! error TS2345: Argument of type '{ bar: T["bar"]; }' is not assignable to parameter of type 'GenericMap<ExtendedConstraint>'.
+!!! error TS2345:   Property 'foo' is missing in type '{ bar: T["bar"]; }' but required in type 'GenericMap<ExtendedConstraint>'.
+!!! related TS2728 tests/cases/compiler/genericMappedTypeAcceptsInferredObjectType.ts:2:2: 'foo' is declared here.
+    	required<ExtendedConstraint>({ foo, bar }) // no error as { foo: T['foo'], bar: T['bar'] } <: GenericMap<ExtendedConstraint>
+    	required<T>({ foo, bar }) // error as { foo: T['foo'], bar: T['bar'] } /<: GenericMap<T> because other properties may be missing
+    	            ~~~~~~~~~~~~
+!!! error TS2345: Argument of type '{ foo: T["foo"]; bar: T["bar"]; }' is not assignable to parameter of type 'GenericMap<T>'.
+    	optional<T>({}) // no error as {} <: OptionalGenericMap<T>
+    	optional<T>({ foo }) // no error as { foo: T['foo'] } <: OptionalGenericMap<T>
+    	            ~~~~~~~

```

Comment:
Looks like this case is not handled yet, even though it should be. I think it's because you're currently relying on the "constraint" type to do too much work; you should not need to look at the constraint for the target at all, it's rarely useful for assignability to a generic type.

Reply:
Yeah, that's what I mentioned in https://github.com/microsoft/TypeScript/pull/42382#discussion_r566676535.